### PR TITLE
V1.3 Filter by Housing Type Preference

### LINKED ISSUE #168: Filter by housing type

### LINKED ISSUE #173: Add testing for filter by housing type

### PR DESCRIPTION
**Pull request by jyztintan**
Fix #168 fix #173 

# Description
This PR introduces the capability to filter entities based on housing type, enhancing the application's flexibility and user experience in managing housing-related data. The following key changes have been made:

# Key Changes
 - Creation of HousingTypeMatchPredicate that checks if 2 `HousingType`s are equal
 - Refactoring TagTest and RemarkTest to the model.person test 
 - Adding test for filtering by housing type
 
# Motivation
The ability to filter by housing type is a directly relevant feature for our real estate agents target audience. This functionality aims to provide users with more precise control over the data they work with.

### CODE REVIEW

File: src/main/java/seedu/realodex/model/person/predicates/HousingTypeMatchPredicate.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import seedu.realodex.commons.util.ToStringBuilder;
+import seedu.realodex.model.person.HousingType;
+import seedu.realodex.model.person.Person;
+
+
+/**
+ * Tests that a {@code Person}'s preferred HousingType equals the specified HousingType.
+ */
+public class HousingTypeMatchPredicate implements Predicate<Person> {
+    private final HousingType housingType;
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/predicates/HousingTypeMatchPredicate.java:14]: 👍

File: src/main/java/seedu/realodex/model/person/predicates/PredicateProducer.java [hunk 2]
```diff
@@ -90,5 +93,26 @@ public Predicate<Person> createMatchTagsPredicate(List<String> tagStrings) {
...
+     *
+     * @param strings The string from which HousingType is represented. This HousingType is then used
+     *               in the predicate to check against a person's preferred HousingType.
+     * @return A {@code Predicate<Person>} that tests whether a person's preferred HousingType equals
+     *         the HousingType represented by the provided string. The predicate returns {@code true}
+     *         if the person's preferred HousingType equals the specified HousingType, and {@code false} otherwise.
+     */
+    public Predicate<Person> createHousingTypeMatchPredicate(List<String> strings) {
+        try {
+            String housingTypeString = strings.get(strings.size() - 1);
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/model/person/predicates/PredicateProducer.java:109]: 👍

File: src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java [hunk 3]
```diff
@@ -150,6 +152,51 @@ public void execute_bothTags_personsWithBothTagsFound() {
         assertEquals(Collections.singletonList(BENSON), model.getFilteredPersonList());
     }
 
+    @Test
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java:155]: Nice work on the tests

File: src/test/java/seedu/realodex/model/person/predicate/PredicateProducerTest.java [hunk 4]
```diff
@@ -66,19 +65,28 @@ void createPredicate_validMultipleTagPrefixes_createsCorrectPredicate() throws P
...
+
+        assertThrows(AssertionError.class, () -> predicateProducer.createMatchTagsPredicate(keyphrases));
     }
 
     @Test
-    void createMatchTagPredicate_invalidTagString_assertionErrorWhenInvalidPrefix() {
+    void createHousingTypeMatchPredicate_validHousingTypeStrings_createsCorrectPredicate() {
         PredicateProducer predicateProducer = new PredicateProducer();
-        List<String> keyphrases = List.of("customer");
+        List<String> keyphrases = List.of("hdb");
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/model/person/predicate/PredicateProducerTest.java:78]: 👍

File: src/test/java/seedu/realodex/model/person/predicate/HousingTypeMatchPredicateTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        // Housing Type matches Good Class Bungalow
+        predicate = new HousingTypeMatchPredicate(new HousingType("Good Class Bungalow"));
+        assertFalse(predicate.test(new PersonBuilder().withHousingType("Landed Property").build()));
+
+        // Housing Type matches Landed Property
+        predicate = new HousingTypeMatchPredicate(new HousingType("Landed Property"));
+        assertFalse(predicate.test(new PersonBuilder().withHousingType("hdb").build()));
+
+    }
+
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/model/person/predicate/HousingTypeMatchPredicateTest.java:83]: 👍

**Review by appleraincoat — APPROVED** [review]
LGTM! Nice work on such short notice

File: src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java [hunk 6]
```diff
@@ -50,7 +52,9 @@ public FilterCommand parse(String args) throws ParseException {
     private Predicate<Person> createPredicateForPrefix(Prefix presentPrefix, List<String> keyphrases)
             throws ParseException {
         checkValidNameIfApplicable(presentPrefix, keyphrases);
+        checkValidRemarkIfApplicable(presentPrefix, keyphrases);
         checkValidTagsIfApplicable(presentPrefix, keyphrases);
+        checkValidHousingTypeIfApplicable(presentPrefix, keyphrases);
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:57]: 👍

File: src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java [hunk 7]
```diff
@@ -130,6 +134,24 @@ private void checkValidNameIfApplicable(Prefix presentPrefix, List<String> keyph
...
+    private void checkValidRemarkIfApplicable(Prefix presentPrefix, List<String> keyphrases) throws ParseException {
+        if (!presentPrefix.equals(PREFIX_REMARK)) {
+            return;
+        }
+        String keyphrase = keyphrases.get(keyphrases.size() - 1);
+        if (keyphrase.isEmpty()) {
+            throw new ParseException(MESSAGE_FILTER_EMPTY_REMARK);
+        }
+    }
+
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:154]: Good check for empty remark, do remember to reflect this on UG that empty remark is omitted on purpose as someone may interpret it as trying to list all clients instead

File: src/main/java/seedu/realodex/model/person/predicates/HousingTypeMatchPredicate.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        HousingTypeMatchPredicate otherHousingTypeMatchPredicate = (HousingTypeMatchPredicate) other;
+        return housingType.equals(otherHousingTypeMatchPredicate.housingType);
+    }
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this).add("Housing Type", housingType).toString();
+    }
+
+}
```
Comment (UdhayaShan1) [inline src/main/java/seedu/realodex/model/person/predicates/HousingTypeMatchPredicate.java:46]: 👍

File: src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java [hunk 9]
```diff
@@ -150,6 +152,51 @@ public void execute_bothTags_personsWithBothTagsFound() {
...
+
+    @Test
+    public void execute_gcbHousingType_personsWithHousingTypeGcbFound() {
+        String expectedMessage = String.format(MESSAGE_PERSONS_LISTED_OVERVIEW, 1);
+        HousingType housingType = new HousingType("Good Class Bungalow");
+        HousingTypeMatchPredicate predicate = new HousingTypeMatchPredicate(housingType);
+        FilterCommand command = new FilterCommand(predicate);
+        expectedModel.updateFilteredPersonList(predicate);
+        assertCommandSuccess(command, model, expectedMessage, expectedModel);
+        assertEquals(Collections.singletonList(DANIEL), model.getFilteredPersonList());
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java:197]: 😊

File: src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java [hunk 10]
```diff
@@ -150,6 +152,51 @@ public void execute_bothTags_personsWithBothTagsFound() {
...
+        assertCommandSuccess(command, model, expectedMessage, expectedModel);
+        assertTrue(model.getFilteredPersonList().containsAll(Arrays.asList(ALICE, ELLE, FIONA)));
+    }
+
+
+    @Test
+    public void execute_condominiumHousingType_personsWithHousingTypeCondominiumFound() {
+        String expectedMessage = String.format(MESSAGE_PERSONS_LISTED_OVERVIEW, 2);
+        HousingType housingType = new HousingType("condominium");
+        HousingTypeMatchPredicate predicate = new HousingTypeMatchPredicate(housingType);
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java:171]: Thanks for comprehensive testing

File: src/test/java/seedu/realodex/logic/parser/FilterCommandParserTest.java [hunk 11]
```diff
@@ -86,6 +93,19 @@ void parse_invalidTagFormat_throwsParseException() {
         assertThrows(IllegalArgumentException.class, () -> new Tag(userInput));
     }
 
+    @Test
+    void parse_validArgsWithHousingType_returnsFilterCommand() {
+        String userInput = " h/hdb";
+        HousingType housingType = new HousingType("hdb");
+        FilterCommand expectedCommand = new FilterCommand(new HousingTypeMatchPredicate(housingType));
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/logic/parser/FilterCommandParserTest.java:102]: nice!

File: src/test/java/seedu/realodex/model/person/predicate/HousingTypeMatchPredicateTest.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    @Test
+    public void test_housingTypeDoesNotMatch_returnsFalse() {
+
+        // Housing Type does not match HDB
+        HousingTypeMatchPredicate predicate = new HousingTypeMatchPredicate(new HousingType("hdb"));
+        assertFalse(predicate.test(new PersonBuilder().withHousingType("Condominium").build()));
+
```
Comment (UdhayaShan1) [inline src/test/java/seedu/realodex/model/person/predicate/HousingTypeMatchPredicateTest.java:69]: Good job!

**Review by UdhayaShan1 — CHANGES_REQUESTED** [review]
LGTM, just need to update the filter messages

**Review by UdhayaShan1 — APPROVED** [review]

File: src/main/java/seedu/realodex/logic/commands/FilterCommand.java [hunk 13]
```diff
@@ -22,25 +23,28 @@ public class FilterCommand extends Command {
...
 
     public static final String MESSAGE_FILTER_HELP = "Filter Command: "
-            + "Filters clients by ONE specified field (name, remark, tag)"
+            + "Filters clients by ONE specified field (name, remark, tag, housing type)"
             + "with the specified keyphrase (non-empty, case-insensitive) "
             + "and displays them as a list with index numbers.\n"
-            + "Format: filter [n/KEYPHRASE] [r/KEYPHRASE] [t/TAG]\n"
-            + "Example: filter n/Jus\n";
+            + "Format: filter [n/KEYPHRASE] [r/KEYPHRASE] [t/TAG] [h/HOUSING_TYPE]\n"
+            + "Example: " + "filter n/alice tan";
```
Comment (wapisai) [inline src/main/java/seedu/realodex/logic/commands/FilterCommand.java:42]: Thankyour for adding housetype! 🆙 🆙 🆙 🆙 🆙 🆙 🆙

File: src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java (hunk 7 above)
Comment (wapisai) [reply src/main/java/seedu/realodex/logic/parser/FilterCommandParser.java:154]: Agreed!

File: src/test/data/JsonSerializableRealodexTest/typicalPersonsRealodex.json [hunk 14]
```diff
@@ -19,7 +19,7 @@
     "address" : "311, Clementi Ave 2, #02-25",
     "family" : "4",
     "tags" : [ "seller", "buyer" ],
-    "housingType" : "HDB",
+    "housingType" : "Condominium",
```
Comment (wapisai) [inline src/test/data/JsonSerializableRealodexTest/typicalPersonsRealodex.json:22]: Thanks for changing up the housing types!

File: src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java (hunk 9 above)
Comment (wapisai) [inline src/test/java/seedu/realodex/logic/commands/FilterCommandTest.java:197]: thanks for testing out all the housing types

File: src/test/java/seedu/realodex/logic/parser/FilterCommandParserTest.java [hunk 15]
```diff
@@ -86,6 +93,19 @@ void parse_invalidTagFormat_throwsParseException() {
...
+
+    @Test
+    void parse_invalidArgsWithHousingType_throwsParseException() {
+        String userInput = " h/hdbb";
+        assertParseFailure(parser, userInput, HousingType.MESSAGE_CONSTRAINTS);
+    }
```
Comment (wapisai) [inline src/test/java/seedu/realodex/logic/parser/FilterCommandParserTest.java:108]: Could consider testing more cases in the future!

File: src/test/java/seedu/realodex/testutil/TypicalPersons.java [hunk 16]
```diff
@@ -47,21 +47,21 @@ public class TypicalPersons {
             .withAddress("311, Clementi Ave 2, #02-25")
             .withEmail("johnd@example.com").withPhone("98765432")
             .withFamily("4")
-            .withHousingType("HDB")
+            .withHousingType("Condominium")
```
Comment (wapisai) [inline src/test/java/seedu/realodex/testutil/TypicalPersons.java:50]: Thanks for changing it up!

**Review by wapisai — COMMENTED** [review]
Good job!

**Review by 4llysa — APPROVED** [review]
LGTM!

**Comment by 4llysa** [comment]
Thanks for fixing the merge conflicts. LGTM!

**Review by appleraincoat — APPROVED** [review]
LGTM

**Review by UdhayaShan1 — APPROVED** [review]
LGTM

**Comment by UdhayaShan1** [comment]
Merged

**Pull request merged**